src/Build/BackEnd/Components/Scheduler/Scheduler.cs

```diff
@@ -1781,6 +1786,66 @@ private ScheduleResponse TrySatisfyRequestFromCache(int nodeForResults, BuildReq
             return null;
         }
 
+        private bool CacheMissEnforcement(int nodeForResults, BuildRequest request, List<ScheduleResponse> responses)
+        {
+            var isIsolatedBuild = _componentHost.BuildParameters.IsolateProjects;
+
+            // do not check root requests as nothing depends on them
+            if (!request.IsRootRequest && isIsolatedBuild)
+            {
+                var configCache = (IConfigCache)_componentHost.GetComponent(BuildComponentType.ConfigCache);
+                var requestConfig = configCache[request.ConfigurationId];
+                
+                // Need the parent request. But the parent / child relationship is sometimes formed after the cache is queried, so get creative
+                var parentRequest = _schedulingData.BlockedRequests.FirstOrDefault(r => r.BuildRequest.GlobalRequestId == request.ParentGlobalRequestId);
+                if (parentRequest == null)
+                {
+                    // the parent might still be in executing requests because the scheduler might not have had a chance to mark it as blocked yet
+                    parentRequest = _schedulingData.ExecutingRequests.FirstOrDefault(r => r.BuildRequest.GlobalRequestId == request.ParentGlobalRequestId);
+                }
+
+                ErrorUtilities.VerifyThrowInternalNull(parentRequest, nameof(parentRequest));
+                ErrorUtilities.VerifyThrow(configCache.HasConfiguration(parentRequest.BuildRequest.ConfigurationId), "All non root requests should have a parent with a loaded configuration");
+
+                var parentConfig = configCache[parentRequest.BuildRequest.ConfigurationId];
+
+                // allow self references (project calling the msbuild task on itself, potentially with different global properties)
```

Comment:
Something I just realized about self-references, what if two different nodes for the same project call into each other? IE, debug node calling into the retail node? Maybe that's an edge case that we don't particularly care about though since it doesn't make much sense anyway...

Reply:
That should be legal under isolated builds right?

Reply:
The idea was to support multitargeting builds, so one project calling into itself. But now that we support nodes with different global props (and multiple entry points), we could have two separate nodes with different global properties, like debug and retail, which in cloudbuild would build on different machines, so semantically we should not allow them to call into each other willy nilly.

However in practice, this seems pretty edge-casey and not breaking as much as not being "correct".

Reply:
Ah, I understand now. Detecting this situation is tricky, because there's no way that I can see to differentiate between two static graph nodes of the same project with different global props and the multi targeting scenario (where you also have the same project calling into itself with different global properties). The difference is that the former would specify the relationship in ProjectReference, but if it misses to do so, there's no way to detect it unless we do more aggressive static analysis over the msbuild task patterns to only whitelist the multi targeting pattern.

Though with our current approach it might still work out. If A -> A' (A' is the same project but different global props), and if they get assigned to different machines, executing A' would lead to its result cache getting serialized and sent over to the machine where A builds. So A would get a cache with the results of A' in it.

During a non distributed in-msbuild graph build you'd get the same result, the cached results for A' would be there when A needs to build.

The problem arises when A does not specify A' in its ProjectReference item. Because the edge is not there, A' does not get built before A. So under isolated builds, it should theoretically fail whereas with this implementation it will allow A to build A' as part of its build.

Reply:
The only thing I can think of to actually solve this would be to treat the global property `TargetFramework` specially here, which . . . ain't great.